CORE-10604 Schema Registry AuthZ: Extend existing ACLs

pgellert opened the pull request:
Introduce the concept of a `subsystem` into ACLs, so that `kafka` resources and `schema_registry` resources can operate independently.

ACLs for `kafka` resources are managed by the Kafka API, and operate on Kafka resources.
ACLs for `schema_registry` resources shall be managed through a separate API.

It is important that the two subsystems do not interact with each other.

This includes a rework of how `acl_binding_filter`'s get serialized to allow for evolving the serialization format. For implementation details, see the commit message.

Replaces https://github.com/redpanda-data/redpanda/pull/26390
Fixes https://redpandadata.atlassian.net/browse/CORE-10604

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

BenPope committed:
schema_registry/authz: Introduce global_resource

Signed-off-by: Ben Pope <ben@redpanda.com>

BenPope committed:
security/acl: Introduce resource_patter_filter::resource_subsystem

ACLs currently assume everything is a kafka resource.

Introduce a mechanism to distinguish kafka resources and
schema registry resources; they operate independently.

Signed-off-by: Ben Pope <ben@redpanda.com>
(cherry picked from commit 8c1815a6359af0e47e8e745117689fbf611f14bf)

BenPope committed:
security: Add support for SR resources.

Add support for two new resources:
* ppsr::subject
* ppsr::global_resource

Signed-off-by: Ben Pope <ben@redpanda.com>
(cherry picked from commit 8620399396ad15534f4722714eec4720a710c25d)

pgellert committed:
security/acl: envelope resource_pattern_filter

resource_pattern_filter was previously serde serialized using the
low-level read_nested and write methods. This meant that it was
serialized without an envelope. This makes it difficult to extend it
with new fields in a backwards compatible way.

To allow adding new fields into it, we introduce a version v1 of it,
that uses serde_fields to serialize it with an envelope header as
expected. However, v1 is not compatible with v0.

We rely on the serde of the containing type to double write the field
and allow switching over to using v1.
See the `acl_binding_filter::serde_write` for implementation details of
this switchover.

pgellert committed:
dt: test ACL operations during upgrade

pgellert committed:
serde/variant: restrict constructor template

Only enable the constructor taking universal references for values that
can be used to construct the underlying variant type. Previously, the
constructor template was considered during overload resolution for any
type, leading to hard-to-debug template instantiation errors when
unrelated types couldn't actually construct the variant.

The fix adds `std::is_constructible_v<type, T>` to the requires clause
to prevent template instantiation for incompatible types.

Example error before fix:

```
bazel-out/k8-fastbuild/bin/src/v/serde/_virtual_includes/variant/serde/rw/variant.h:64:9: error: no matching constructor for initialization of 'std::variant<pattern_type, pattern_match>'
   64 |       : type(std::forward<T>(t)){};
      |         ^    ~~~~~~~~~~~~~~~~~~
bazel-out/k8-fastbuild/bin/src/v/json/_virtual_includes/json/json/json.h:100:28: note: in instantiation of function template specialization 'serde::variant<security::pattern_type, security::resource_pattern_filter::pattern_match>::variant<const seastar::basic_sstring<char, unsigned int, 15> &>' requested here
  100 |         rjson_serialize(w, e);
      |                            ^
bazel-out/k8-fastbuild/bin/src/v/compat/_virtual_includes/compat/compat/json.h:390:5: note: in instantiation of function template specialization 'json::rjson_serialize<rapidjson::GenericStringBuffer<rapidjson::UTF8<>, json::throwing_allocator>, seastar::basic_sstring<char, unsigned int, 15>, std::allocator<seastar::basic_sstring<char, unsigned int, 15>>>' requested here
  390 |     rjson_serialize(w, b.mount_paths);
      |     ^
```

pgellert commented:
[force-push](https://github.com/redpanda-data/redpanda/compare/38e373f1ddeb9757fd6cf605b88cafa2281ef369..811756bcd1d97fbca0a77e47ac0bfe08cda61cc7): fix compilation errors of src/v/compat (see https://github.com/redpanda-data/redpanda/pull/26466/commits/af04eea25490241c25ba496e57b38f266ec0297a)

dotnwat commented on src/v/security/acl.h:
```diff
@@ -403,12 +407,15 @@ class acl_binding
  */
 class resource_pattern_filter
   : public serde::envelope<
       resource_pattern_filter,
-      serde::version<0>,
-      serde::compat_version<0>> {
+      serde::version<1>,
+      serde::compat_version<1>> {
```
I haven't fully reviewed this yet, but my initial thought is:

Why are we trying to force the envelope back into existence (and doing this like waiting for V0 to no be supported) as opposed to just accepting that `resource_pattern_filter` will always use a non-standard encoding?

Maybe it won't work, but what I had in mind was something like this:

```cpp
struct resource_pattern_filter {
    // these 3 fields are forever written without any versioning
    std::optional<resource_type> _resource;
    std::optional<ss::sstring> _name;
    std::optional<pattern_filter_type> _pattern;

    struct new_stuff : envelope<> {
      ...
    };
};
```

Assuming that old code will skip extra data on the wire (which seems true based on how this PR looks), then

```
write() {
  raw write resource, name, pattern
  envelope write new_stuff
}

read() {
   raw read the three fields
   if extra data, envelope decode new_stuff
}
```

anyway, i don't think there is a "correct" implementation here, i just want to highlight that we don't need to accept undesirable complexity / fragility just to achieve the reintroduction of the envelope.

pgellert replied:
This wouldn't work because we can't determine how much extra data belongs to `resource_pattern_filter` since it lacks its own envelope. The `bytes_left_limit` parameter in `read_nested` refers to bytes remaining in the outer envelope (`acl_binding_filter`), not the current struct. Ref: https://github.com/redpanda-data/redpanda/blob/f212ce0f51964f15bf10121cb8b39dd2ad4ce707/src/v/serde/rw/envelope.h#L69-L85

Additionally, old readers would fail to parse `acl_binding_filter` correctly - they'd interpret `new_stuff` as the start of the next field (`acl_entry_filter _acl`) rather than part of `resource_pattern_filter`.

An alternative would be adding a `resource_pattern_filter_extension : serde::envelope<>` field at the end of `acl_binding_filter`. Old readers would ignore it, and we'd avoid duplicating existing fields. However, this splits `resource_pattern_filter` data across non-contiguous blocks and makes the structure harder to reuse outside of `acl_binding_filter` contexts.

Therefore, deprecating the old field and introducing a new enveloped field seems like the cleanest approach to me.

dotnwat replied:
> Assuming that old code will skip extra data on the wire (which seems true based on how this PR looks)

> This wouldn't work because we can't determine how much extra data belongs to resource_pattern_filter since it lacks its own envelope. 

Ahh, I thought that your new `serde_write` was for the resource pattern, but it was for the acl binding filter, so that makes sense.

Can we add unit tests for the backwards/forward/upgrade test scenarios? You can make a copy of the old code into special namespaces that indicate how they are going to be used.

pgellert replied:
Makes sense. I've added these now: https://github.com/redpanda-data/redpanda/pull/26466#issuecomment-2980171907

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--22ec1007-03d6-4dd8-8b85-3a4d9cec0b80-->
<!--TEST RESULTS-->
<!--505e87fb-f69c-400e-9f5f-4d277dde01f2-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67335">build#67335</a></summary>


| test_class                 | test_method                                       | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:--------------------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest  | test_read_chunks                                  |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67335#01976a83-220f-4143-b5a6-1bbe50167d83 | FLAKY         | 19/21    | upstream reliability is '97.22222222222221'. current run reliability is '90.47619047619048'. drift is 6.74603 and the allowed drift is set to 50. The test should PASS |
| PartitionReassignmentsTest | test_add_partitions_with_inprogress_reassignments |                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67335#01976a83-220f-4de6-a963-35ad8931dffa | FLAKY         | 19/21    | upstream reliability is '92.15384615384615'. current run reliability is '90.47619047619048'. drift is 1.67766 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67462">build#67462</a></summary>


| test_class      | test_method             | test_arguments     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------|:------------------------|:-------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MaintenanceTest | test_maintenance_sticky | {"use_rpk": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67462#01977e1d-a554-4bf0-a6ae-ee06752bcc76 | FLAKY         | 20/21    | upstream reliability is '95.96273291925466'. current run reliability is '95.23809523809523'. drift is 0.72464 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

mmaslankaprv commented:
lgtm

pgellert committed:
security/test: fwd/bwd acl serde tests

Add serde compatibility tests for the `acl_binding_filter` to document
and test the expected behaviour during the evolution from v0 through v2.

pgellert commented:
[pushed](https://github.com/redpanda-data/redpanda/pull/26466/commits/9f8e48b08e23abe4331b1880e3dadf31bf5fd059): unit testing of serde compatibility across versions
